### PR TITLE
fix: wrong usage of Control generic type in FormArray for FormBuilder…

### PR DESCRIPTION
**Pull request by jogelin**
… and for ValuesOf

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/reactive-forms/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ x ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #137 

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: libs/reactive-forms/src/lib/types.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export type ValuesOf<T extends ControlsOf<any>> = {
   ? R
   : NonUndefined<T[K]> extends FormGroup<infer R>
   ? ValuesOf<R>
-  : NonUndefined<T[K]> extends FormArray<infer R> ? R[] : NonUndefined<T[K]>;
+  : NonUndefined<T[K]> extends FormArray<infer R, infer C> ? R[] : NonUndefined<T[K]>;
```
Comment (NetanelBasal) [inline libs/reactive-forms/src/lib/types.ts:22]: infer C is redudant no?

Comment (jogelin) [reply libs/reactive-forms/src/lib/types.ts:22]: I know this is strange but this is the only way I found to have the type matching:
- `NonUndefined<T[K]> extends FormArray<infer R>` match `FormArray<R>` and return `R[]`
- `NonUndefined<T[K]> extends FormArray<infer R>` doesn't match `FormArray<R, FormControl<R>>` and return `NonUndefined<T[K]>`
- `NonUndefined<T[K]> extends FormArray<infer R, infer C>` match `FormArray<R, FormControl<R>>` and return `R[]` 

TBH I am not sure about this solution. I just want `ValuesOf` to work when specifying a `Control` type

**Comment by jogelin** [comment]
@NetanelBasal I forgot to say that we are working with v3.1.4 because we are not yet compatible with angular 13. 
This is the reason why we are doing this migration. Would it be possible to also make a release in v3 too?
Or the v4 is also compatible with angular v12 (didn't try)?

**Comment by NetanelBasal** [comment]
I've created a [v12](https://github.com/ngneat/reactive-forms/tree/ng-12) branch. You should checkout and merge all the relevant changes to this branch. Then, I'll make a release for this.

**Comment by jogelin** [comment]
> I've created a [v12](https://github.com/ngneat/reactive-forms/tree/ng-12) branch. You should checkout and merge all the relevant changes to this branch. Then, I'll make a release for this.

I tested with the v4 and it seems backward compatible with angular v12 so you can just make a v4 release ;) 
Waiting for it :p

**Comment by NetanelBasal** [comment]
It's not, it's Ivy distribution.

**Comment by NetanelBasal** [comment]
I have already released it.

**Comment by jogelin** [comment]
> It's not, it's Ivy distribution.

correct..this is because I am using Ivy already then ;)

**Comment by jogelin** [comment]
> I have already released it.

it works on our side, thanks ;)

**Pull request merged**